### DISCUSSION TITLE
train with train_text_to_image_sdxl.py  CUDA out of memory

### DISCUSSION THREAD

**Discussion by MMY1994** [huggingface/diffusers#7870] [General]
![17149912736562](https://github.com/huggingface/diffusers/assets/33061770/5aec6621-56f9-4133-a042-5444745cbb21)

This is my shell script

export MODEL_NAME="./model_base/sd_xl_base"
export VAE_NAME="./model_base/sdxl-vae-fp16-fix"

accelerate launch \
  --config_file /root/.cache/huggingface/accelerate/default_config.yaml \
  --num_cpu_threads_per_process=8 \
  ./diffusers/examples/text_to_image/train_text_to_image_sdxl_mmy.py \
  --pretrained_model_name_or_path=$MODEL_NAME \
  --pretrained_vae_model_name_or_path=$VAE_NAME \
  --train_data_dir="./fineturn/fineturn.py" \
  --output_dir="./sdxl-dreamshape-fineturn-model/" \
  --enable_xformers_memory_efficient_attention \
  --resolution=1024 --center_crop --random_flip\
  --proportion_empty_prompts=0.2 \
  --train_batch_size=1 \
  --mixed_precision="fp16" \
  --gradient_accumulation_steps=4 --gradient_checkpointing \
  --max_train_steps=5000 \
  --use_8bit_adam \
  --learning_rate=1e-07 \
  --lr_scheduler="constant" --lr_warmup_steps=0 \
  --validation_prompt="a young man climbing the ladder, high quality" --validation_epochs 1 \
  --checkpointing_steps=200

I Use A800 40GB and --enable_xformers_memory_efficient_attention, --mixed_precision="fp16".

Even when I use 512 resolution and batchsize=1, this error still happen. i don't think it's normal

**Comment by tolgacangoz** [comment]
Could you share the output of `diffusers-cli env` command? I see that you added a suffix to the file name and `    main(args)` is at 1333. line, is there any change in the file in this discussion's context? I wonder if `--num_cpu_threads_per_process=8` tries to allocate too much data? Also, what happens when you remove `--enable_xformers_memory_efficient_attention` if your PyTorch>=2.0.

**Reply by MMY1994** [reply]
this file just copy from train_text_to_image_sdxl.py no change
diffusers-cli env is
![17150033958932](https://github.com/huggingface/diffusers/assets/33061770/2872892c-df44-4a74-a9ef-4196c4ca8faa)

**Reply by MMY1994** [reply]
I found that lowering the resolution didn't work, either 256, 512 or 1024 would report an error. So its core problem is probably not the size of GPU memory

**Reply by tolgacangoz** [reply]
Did you try to lower `--num_cpu_threads_per_process=8` and remove `--enable_xformers_memory_efficient_attention`?

**Reply by MMY1994** [reply]
Yes, I've tried --num_cpu_threads_per_process=1 or 2 and remove --enable_xformers_memory_efficient_attention. This error still report.

**Reply by tolgacangoz** [reply]
`--gradient_accumulation_steps=8`?
Also, what happens if you choose torch.compile when configuring accelerate? To benefit torch.compile most, could you upgrade `torch`?

**Reply by MMY1994** [reply]
I tried it, too. when --gradient_accumulation_steps=8, this error will report at step 8, and --gradient_accumulation_steps=1, this error will report at step 1

**Reply by MMY1994** [reply]
Even if I don't use these methods to speed up or reduce GPU memory, memory should not be insufficient when batchsize=1 and the resolution is 256

**Reply by MMY1994** [reply]
I found no matter how to adjust the resolution,256,512,or 1024, model_input'shape will always be[b, 4, 64, 64]。But examples["pixel_values"] is changed
![17150737312758](https://github.com/huggingface/diffusers/assets/33061770/cf16fe10-9c97-4eda-9047-4dfb69b46f61)

**Reply by tolgacangoz** [reply]
SD-XL is a Latent Diffusion Model, so it works on a lower dimensional latent space while noising/denoising.

**Reply by MMY1994** [reply]
But when the input resolution is 1024*1024, its latent space dimension should be 128*128? I adjust resolution from 1024 to 256, its latent space dimension hasn't changed

**Reply by MMY1994** [reply]
Sorry to ask , have you try the input resolution of 1024, batch of 1, in the case of 40G GPU memory training

**Reply by tolgacangoz** [reply]
No, since I am GPU-poor :smile:.
[Some other one also mentioned this situation before.](https://github.com/huggingface/diffusers/issues/7114) It seems that 512x512 requires 40+ GB.
[Maybe `accelerate`'s `deepspeed` options can be tried.](https://huggingface.co/docs/accelerate/main/en/usage_guides/deepspeed)

So, it seems that resolution is always processed as 512x512? Could you trace the root cause of this?

**Reply by tolgacangoz** [reply]
I have just checked `model_input`'s shape. It is 32x32 when `resolution` is 256. Could you double-check?

**Reply by tolgacangoz** [reply]
What happens if you set `--gradient_accumulation_steps=1`? In this case, it won't accumulate gradients, it directly makes the update and clears them AFAIU.

**Reply by MMY1994** [reply]
I checked it again, and it was weird
![17151365252769](https://github.com/huggingface/diffusers/assets/33061770/089fd476-152c-4cbd-ac0d-2e897d3f3f99)
![1715136623293](https://github.com/huggingface/diffusers/assets/33061770/7cf50893-f6af-43ac-9666-33e1e6f23636)
Is the data where the cache has been called

**Reply by MMY1994** [reply]
Yes, it's still produce OOM

**Reply by tolgacangoz** [reply]
Good news! I am able to train 1024x1024 with a 24GB GPU!
Replace `AdamW8bit` with [`PagedAdamW8bit`](https://huggingface.co/docs/bitsandbytes/main/en/explanations/optimizers#paged-optimizers). But I cannot evaluate its speed.

**Reply by MMY1994** [reply]
Now i use PageAdamW8bit and single GPU, still have 2 problem
1. when I set batchsize to 1 or 8,  the GPU memory used is about 40G. There is no difference. Set to 20, it produce OOM
2. No matter how I change the resolution, the model_input's shape still is [b, 4, 64, 64].
It feels like these two parameters don't work at all.
I installed diffusers by python setup.py develop, what effect will this have?

**Reply by MMY1994** [reply]
I found it when i use train_dataset.map(......), the second error will report.
![17151716104905](https://github.com/huggingface/diffusers/assets/33061770/4e1a2be6-4d30-43f4-9649-4f525694e174)
But when i use train_dataset like train_text_to_image_lora_sdxl.py, the model_input's shape is normal

**Reply by tolgacangoz** [reply]
1. `PagedAdamW8bit` uses system RAM when OOM seems to occur in GPU, so it is normal to use all GPU vRAM -it uses all GPU vRAM + less system RAM with a batch size of 1, and uses all GPU vRAM + more system RAM with a batch size of 8, AFAIU. For the further batch sizes, [the `deepspeed` library](https://huggingface.co/docs/accelerate/main/en/usage_guides/deepspeed) could be researched.
2. I couldn't reproduce this with the setup below. Maybe, your `torchvision` and/or `datasets` are problematic; upgrade them?


~25 GB system RAM used at peak, 24 GB vRAM NVIDIA L4.
Command:
```sh
accelerate launch
--num_cpu_threads_per_process=1
./diffusers/examples/text_to_image/train_text_to_image_sdxl.py
--pretrained_model_name_or_path="stabilityai/stable-diffusion-xl-base-1.0"
--pretrained_vae_model_name_or_path="madebyollin/sdxl-vae-fp16-fix"
--train_data_dir="YaYaB/onepiece-blip-captions"
--output_dir="./sdxl-dreamshape-fineturn-model/"
--resolution=1024 --center_crop --random_flip
--proportion_empty_prompts=0.2
--train_batch_size=1
--mixed_precision="fp16"
--gradient_checkpointing
--max_train_steps=10
--use_8bit_adam
--learning_rate=1e-07
--lr_scheduler="constant" --lr_warmup_steps=0
--validation_prompt="a young man climbing the ladder, high quality" --validation_epochs 1
--checkpointing_steps=200
```
with this modification in the `examples/text_to_image/train_text_to_image_sdxl.py` at the 790. line:
```diff
-        optimizer_class = bnb.optim.AdamW8bit
+        optimizer_class = bnb.optim.PagedAdamW8bit
```
`accelerate env`:
```sh
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: NO
        - mixed_precision: fp16
        - use_cpu: False
        - debug: False
        - num_processes: 1
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: all
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
```
`diffusers-cli env`:
```sh
- `diffusers` version: 0.28.0.dev0
- Platform: Linux-5.15.0-1058-aws-x86_64-with-glibc2.31
- Python version: 3.10.10
- PyTorch version (GPU?): 2.4.0.dev20240507+cu121 (True)
- Huggingface_hub version: 0.23.0
- Transformers version: 4.40.2
- Accelerate version: 0.30.0
- xFormers version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>
```

```sh
python -c "import bitsandbytes as bnb, torchvision as tv, datasets as ds; print('bnb:', bnb.__version__); print('torchvision:', tv.__version__); print('datasets:', ds.__version__);"
bnb: 0.43.1
torchvision: 0.19.0.dev20240507+cu121
datasets: 2.19.1
```

**Reply by MMY1994** [reply]
Thank you, I will try it

**Reply by MMY1994** [reply]
After try it, GPU memery problem seems to be related to the torch version(my torch version is 2.0.1)。accelerate config's distributed_type set to "NO" is norml( single GPU about 25GB), but set to "MULTI_GPU"is about 38GB. it reported the Info "torch.nn.parallel.distributed - Reducer buckets have been rebuilt in this iteration".
I will try upgrade them

**Reply by MMY1994** [reply]
I'm sorry to interrupt, but have you tried single machine multi GPU and distribution, like this

accelerate env:

- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: fp16
        - use_cpu: False
        - debug: False
        - num_processes: 8
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: all
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []

According to the results of my experiment, single GPU is normal, but multi-GPUs still takes up too much GPU memory

**Reply by tolgacangoz** [reply]
AFAIU, [`PagedAdamW8bit`](https://huggingface.co/docs/bitsandbytes/main/en/explanations/optimizers#paged-optimizers) uses as much GPU vRAM as possible. If there seems to be a possibility of OOM, then at that point, it starts to utilize CPU RAM. If multi-GPUs have more vRAM than single GPU then multi-GPUs will use more vRAM according to training needs.
Actually, my knowledge about distributed training is very limited. It might be more proper to open a new discussion about that.

**Reply by wzhiyuan2016** [reply]
> I'm sorry to interrupt, but have you tried single machine multi GPU and distribution, like this
> 
> accelerate env:
> 
> * `Accelerate` default config:
>   - compute_environment: LOCAL_MACHINE
>   - distributed_type: MULTI_GPU
>   - mixed_precision: fp16
>   - use_cpu: False
>   - debug: False
>   - num_processes: 8
>   - machine_rank: 0
>   - num_machines: 1
>   - gpu_ids: all
>   - rdzv_backend: static
>   - same_network: True
>   - main_training_function: main
>   - enable_cpu_affinity: False
>   - downcast_bf16: no
>   - tpu_use_cluster: False
>   - tpu_use_sudo: False
>   - tpu_env: []
> 
> According to the results of my experiment, single GPU is normal, but multi-GPUs still takes up too much GPU memory

Hello, did you successfully train in the end? Have you trained your own dataset? How does the trained model generate
I am training my own data and the generation effect is very poor. I don't know why
Some of my discussions are here====>https://github.com/huggingface/diffusers/discussions/10195

**Comment by DuShunpeng** [comment]
> ![17149912736562](https://private-user-images.githubusercontent.com/33061770/328149214-5aec6621-56f9-4133-a042-5444745cbb21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2NDMxOTQsIm5iZiI6MTcxNzY0Mjg5NCwicGF0aCI6Ii8zMzA2MTc3MC8zMjgxNDkyMTQtNWFlYzY2MjEtNTZmOS00MTMzLWEwNDItNTQ0NDc0NWNiYjIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA2VDAzMDEzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgyOGRkZWQ5NzVhZjM2MGUzNjNmYzY5OTEzZTYyMmMyNDhhOTQ0ZGUwYzk4ZDYzYjc2MDAyOTRhZTkzZDhmMDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3rMwvF_z7iC0hZxVD0kMItwbbhwtIjbjMEaXdZ0o8xA)
> 
> This is my shell script
> 
> export MODEL_NAME="./model_base/sd_xl_base" export VAE_NAME="./model_base/sdxl-vae-fp16-fix"
> 
> accelerate launch --config_file /root/.cache/huggingface/accelerate/default_config.yaml --num_cpu_threads_per_process=8 ./diffusers/examples/text_to_image/train_text_to_image_sdxl_mmy.py --pretrained_model_name_or_path=$MODEL_NAME --pretrained_vae_model_name_or_path=$VAE_NAME --train_data_dir="./fineturn/fineturn.py" --output_dir="./sdxl-dreamshape-fineturn-model/" --enable_xformers_memory_efficient_attention --resolution=1024 --center_crop --random_flip --proportion_empty_prompts=0.2 --train_batch_size=1 --mixed_precision="fp16" --gradient_accumulation_steps=4 --gradient_checkpointing --max_train_steps=5000 --use_8bit_adam --learning_rate=1e-07 --lr_scheduler="constant" --lr_warmup_steps=0 --validation_prompt="a young man climbing the ladder, high quality" --validation_epochs 1 --checkpointing_steps=200
> 
> I Use A800 40GB and --enable_xformers_memory_efficient_attention, --mixed_precision="fp16".
> 
> Even when I use 512 resolution and batchsize=1, this error still happen. i don't think it's normal

hi, @MMY1994 :
im not sure that "35GB when setting --train_batch_size=1 and --resolution=1024." is true or not,
but pls try  as below:
1、make sure that " --use_8bit_adam  --enable_xformers_memory_efficient_attention --set_grads_to_none "
2、find "optimizer_class = bnb.optim.AdamW8bit"   =>  "bnb.optim.PagedAdamW8bit " . train_controlnet_sdxl.py or whatever you are using

Good luck!
